File: packages/agent-mesh/sdks/typescript/src/trust.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
import * as fs from 'fs';
import { AgentIdentity } from './identity';
import {
  TrustConfig,
  TrustScore,
  TrustTier,
  TrustVerificationResult,
} from './types';

const DEFAULT_THRESHOLDS = {
  untrusted: 0.0,
  provisional: 0.3,
  trusted: 0.6,
  verified: 0.85,
};

interface AgentTrustState {
  successes: number;
  failures: number;
  score: number;
  lastUpdate: number;
}

/**
 * Manages trust scores for peer agents using a simple Bayesian-inspired model
 * with configurable decay.
 */
export class TrustManager {
  private readonly config: Required<Omit<TrustConfig, 'persistPath'>>;
  private readonly agents: Map<string, AgentTrustState> = new Map();
  private readonly persistPath?: string;

  constructor(config?: TrustConfig) {
    this.config = {
      initialScore: config?.initialScore ?? 0.5,
      decayFactor: config?.decayFactor ?? 0.95,
      thresholds: { ...DEFAULT_THRESHOLDS, ...config?.thresholds },
    };
    this.persistPath = config?.persistPath;
    if (this.persistPath) {
      this.loadFromDisk();
    }
  }

  /** Verify a peer agent's identity and return a trust result. */
  async verifyPeer(
    peerId: string,
    peerIdentity: AgentIdentity,
  ): Promise<TrustVerificationResult> {
    // Verify the identity is self-consistent (DID contains the key fingerprint)
    const challenge = new Uint8Array(
      require('crypto').randomBytes(32),
    );
    const signature = peerIdentity.sign(challenge);
    const verified = peerIdentity.verify(challenge, signature);

    const trustScore = this.getTrustScore(peerId);

    return {
      verified,
      trustScore,
      reason: verified ? undefined : 'Identity verification failed',
    };
  }

  /** Get the current trust score for an agent. */
  getTrustScore(agentId: string): TrustScore {
    const state = this.getOrCreateState(agentId);
    this.applyDecay(state);

    const tier = this.computeTier(state.score);

    return {
      overall: Math.round(state.score * 1000) / 1000,
      dimensions: {
        reliability: this.computeReliability(state),
        consistency: state.score,
      },
      tier,
    };
  }

  /** Record a successful interaction with an agent. */
  recordSuccess(agentId: string, reward: number = 0.05): void {
    const state = this.getOrCreateState(agentId);
    this.applyDecay(state);
    state.successes += 1;
    state.score = Math.min(1, state.score + reward);
    state.lastUpdate = Date.now();
    this.saveToDisk();
  }

  /** Record a failed interaction with an agent. */
  recordFailure(agentId: string, penalty: number = 0.1): void {
    const state = this.getOrCreateState(agentId);
    this.applyDecay(state);
    state.failures += 1;
    state.score = Math.max(0, state.score - penalty);
    state.lastUpdate = Date.now();
    this.saveToDisk();
  }

  // ── Private helpers ──

  private getOrCreateState(agentId: string): AgentTrustState {
    let state = this.agents.get(agentId);
    if (!state) {
      state = {
        successes: 0,
        failures: 0,
        score: this.config.initialScore,
        lastUpdate: Date.now(),
      };
      this.agents.set(agentId, state);
    }
    return state;
  }

  private applyDecay(state: AgentTrustState): void {
    const elapsed = Date.now() - state.lastUpdate;
    const hours = elapsed / (1000 * 60 * 60);
    if (hours >= 1) {
      const decaySteps = Math.floor(hours);
      state.score *= Math.pow(this.config.decayFactor, decaySteps);
      state.lastUpdate = Date.now();
    }
  }

  private computeTier(score: number): TrustTier {
    const t = this.config.thresholds;
    if (score >= t.verified) return 'Verified';
    if (score >= t.trusted) return 'Trusted';
    if (score >= t.provisional) return 'Provisional';
    return 'Untrusted';
  }

  private computeReliability(state: AgentTrustState): number {
    const total = state.successes + state.failures;
    if (total === 0) return this.config.initialScore;
    return Math.round((state.successes / total) * 1000) / 1000;
  }

  private saveToDisk(): void {
    if (!this.persistPath) return;
    try {
      const data: Record<string, AgentTrustState> = {};
      for (const [key, value] of this.agents) {
        data[key] = value;
      }
      fs.writeFileSync(this.persistPath, JSON.stringify(data), 'utf-8');
    } catch {
      // best-effort: ignore write errors
    }
  }

  private loadFromDisk(): void {
    if (!this.persistPath) return;
    try {
      const raw = fs.readFileSync(this.persistPath, 'utf-8');
      const data = JSON.parse(raw) as Record<string, AgentTrustState>;
      for (const [key, value] of Object.entries(data)) {
        this.agents.set(key, value);
      }
    } catch {
      // best-effort: ignore missing or corrupt files
    }
  }
}

File: packages/agent-mesh/sdks/go/trust.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package agentmesh

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sync"
)

// TrustScore represents an agent's current trust standing.
type TrustScore struct {
	Overall    float64            `json:"overall"`
	Dimensions map[string]float64 `json:"dimensions"`
	Tier       string             `json:"tier"`
}

type scoreState struct {
	score        float64
	interactions int
}

type persistedScores struct {
	Scores map[string]*persistedState `json:"scores"`
}

type persistedState struct {
	Score        float64 `json:"score"`
	Interactions int     `json:"interactions"`
}

// TrustManager tracks and updates per-agent trust scores.
type TrustManager struct {
	mu     sync.RWMutex
	config TrustConfig
	scores map[string]*scoreState
}

// NewTrustManager creates a TrustManager with the given config.
func NewTrustManager(config TrustConfig) *TrustManager {
	tm := &TrustManager{
		config: config,
		scores: make(map[string]*scoreState),
	}
	if config.PersistPath != "" {
		_ = tm.loadFromDisk()
	}
	return tm
}

// VerifyPeer verifies a peer's identity and returns the current trust score.
func (tm *TrustManager) VerifyPeer(peerID string, peerIdentity *AgentIdentity) (*TrustVerificationResult, error) {
	verified := peerIdentity != nil && peerIdentity.PublicKey != nil && len(peerIdentity.PublicKey) == 32
	score := tm.GetTrustScore(peerID)
	return &TrustVerificationResult{
		PeerID:   peerID,
		Verified: verified,
		Score:    score,
	}, nil
}

// GetTrustScore returns the current trust score for an agent.
func (tm *TrustManager) GetTrustScore(agentID string) TrustScore {
	tm.mu.RLock()
	defer tm.mu.RUnlock()

	s, ok := tm.scores[agentID]
	if !ok {
		return TrustScore{
			Overall:    tm.config.InitialScore,
			Dimensions: map[string]float64{"reliability": tm.config.InitialScore},
			Tier:       tm.tierFor(tm.config.InitialScore),
		}
	}

	return TrustScore{
		Overall:    s.score,
		Dimensions: map[string]float64{"reliability": s.score},
		Tier:       tm.tierFor(s.score),
	}
}

// RecordSuccess increases an agent's trust score with decay.
func (tm *TrustManager) RecordSuccess(agentID string, reward float64) {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	s := tm.getOrCreate(agentID)
	s.interactions++
	decayed := tm.applyDecay(s.score)
	s.score = math.Min(1.0, decayed+reward*tm.config.RewardFactor)
	_ = tm.saveToDisk()
}

// RecordFailure decreases an agent's trust score with asymmetric penalty.
func (tm *TrustManager) RecordFailure(agentID string, penalty float64) {
	tm.mu.Lock()
	defer tm.mu.Unlock()

	s := tm.getOrCreate(agentID)
	s.interactions++
	decayed := tm.applyDecay(s.score)
	s.score = math.Max(0.0, decayed-penalty*tm.config.PenaltyFactor)
	_ = tm.saveToDisk()
}

func (tm *TrustManager) getOrCreate(agentID string) *scoreState {
	s, ok := tm.scores[agentID]
	if !ok {
		s = &scoreState{score: tm.config.InitialScore}
		tm.scores[agentID] = s
	}
	return s
}

func (tm *TrustManager) applyDecay(score float64) float64 {
	return score * (1.0 - tm.config.DecayRate)
}

func (tm *TrustManager) tierFor(score float64) string {
	switch {
	case score >= tm.config.TierThresholds.High:
		return "high"
	case score >= tm.config.TierThresholds.Medium:
		return "medium"
	default:
		return "low"
	}
}

func (tm *TrustManager) loadFromDisk() error {
	if tm.config.PersistPath == "" {
		return nil
	}
	data, err := os.ReadFile(tm.config.PersistPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("reading trust state: %w", err)
	}
	var persisted persistedScores
	if err := json.Unmarshal(data, &persisted); err != nil {
		return fmt.Errorf("unmarshalling trust state: %w", err)
	}
	for id, ps := range persisted.Scores {
		tm.scores[id] = &scoreState{
			score:        ps.Score,
			interactions: ps.Interactions,
		}
	}
	return nil
}

func (tm *TrustManager) saveToDisk() error {
	if tm.config.PersistPath == "" {
		return nil
	}
	persisted := persistedScores{
		Scores: make(map[string]*persistedState),
	}
	for id, s := range tm.scores {
		persisted.Scores[id] = &persistedState{
			Score:        s.score,
			Interactions: s.interactions,
		}
	}
	data, err := json.MarshalIndent(persisted, "", "  ")
	if err != nil {
		return fmt.Errorf("marshalling trust state: %w", err)
	}
	return os.WriteFile(tm.config.PersistPath, data, 0644)
}
